### ISSUE TITLE
azuread_service_principal not properly provisioned before running its dependencies

### ISSUE THREAD

**Issue by vp777** [hashicorp/terraform-provider-azuread#652]
I am trying to create an application with a service principal using the following config:

```
resource "azuread_application" "main" {
  display_name    = "main-reader"
  sign_in_audience = "AzureADMyOrg"
  fallback_public_client_enabled = true
  
  web {
    implicit_grant {
      access_token_issuance_enabled = true
      id_token_issuance_enabled     = true
    }
  }
}

resource "azuread_service_principal" "main" {
  application_id = "${azuread_application.main.application_id}"
  
  provisioner "local-exec" {
       #the following command is used to grant admin consent for certain permissions
	command = "az ad app permission grant --id ${azuread_service_principal.main.application_id} --api 00000003-0000-0000-c000-000000000000 --scope 'User.Read.All'" #this command fails
	#command = "sleep 60;az ad app permission grant --id ${azuread_service_principal.main.application_id} --api 00000003-0000-0000-c000-000000000000 --scope 'User.Read.All'" #this command runs smoothly

	interpreter = ["powershell", "-c"]
  }
}
```

But while trying to provision this configuration i get the error:
```
│ Error running command 'az ad app permission grant --id 602e4230-8e6b-4cdb-8daf-da9a097d2055 --api 00000003-0000-0000-c000-000000000000 --scope 'User.Read.All'': exit status 1. Output:
│ ERROR: Service principal with appId or objectId '602e4230-8e6b-4cdb-8daf-da9a097d2055' doesn't exist. If '602e4230-8e6b-4cdb-8daf-da9a097d2055' is an appId, make sure an associated
│ service principal is created for the app. To create one, run `az ad sp create --id 602e4230-8e6b-4cdb-8daf-da9a097d2055`
```

But if i add a delay, e.g. sleep 60 before trying to grant a permission then it executes correctly.
It looks like the azuread_service_principal returns prematurely. I have also tried using a null_resource with dependency on the service principal but i had the same result.

**Comment by katbyte** [comment]
Hey @vp777 - Sorry your having this issue but i believe it is a problem with the AAD provider not azurerm so as such i am going to transfer it

**Comment by manicminer** [comment]
Hi @vp777, thanks for reporting this. Handling replication delays like these is challenging because the API caches new objects, making it impossible for us to tell for sure whether an object has finished creating. There is a trick we can do to detect early replication delay, and whilst it does not always work, whilst testing using your example config I was able to go from around 80% failure of the provisioner to around 10% failure - I've added this mitigation in #656.

For additional data points, I added some counts to your config to create 10-20 applications and service principals in parallel, with a provisioner for each instance, and whilst the failure rate was improved slightly I still saw maybe half of them regularly fail. Unfortunately this is the best we can presently do to mitigate this, but we will continue engaging with the service team to see if this can be improved all round.
